libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -0,0 +1,243 @@
+        /// </summary>
+        /// <param name="request">A request object from Restify or Express.</param>
+        /// <param name="response">A response object from Restify or Express.</param>
+        /// <param name="bot">A bot with logic function in the form `async(context) => { ... }`.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task ProcessAsync(HttpRequest request, HttpResponse response, IBot bot, CancellationToken cancellationToken = default)
+        {
+            response.StatusCode = 200;
+            await response.WriteAsync(string.Empty);
+
+            var twilioSignature = request.Headers["x-twilio-signature"];
+
+            var validationUrl = _options.ValidationUrl ?? (request.Headers["x-forwarded-proto"][0] ?? request.Protocol + "://" + request.Host + request.Path);
+
+            var requestValidator = new RequestValidator(_options.AuthToken);
+
+            var bodyStream = new StreamReader(request.Body);
+
+            dynamic payload = bodyStream.ReadToEnd();
+
+            var formattedPayload = payload.Replace("+", "%20");
+
+            var body = QueryStringToDictionary(formattedPayload);
+
+            if (!twilioSignature.Equals(string.Empty) && requestValidator.Validate(validationUrl, body, twilioSignature))
+            {
+                var jsonBody = JsonConvert.SerializeObject(body);
+
+                TwilioEvent twilioEvent = JsonConvert.DeserializeObject<TwilioEvent>(jsonBody);
+
+                var activity = new Activity()
+                {
+                    Id = twilioEvent.MessageSid,
+                    Timestamp = new DateTime(),
+                    ChannelId = "twilio-sms",
+                    Conversation = new ConversationAccount()
+                    {
+                        Id = twilioEvent.From,
+                    },
+                    From = new ChannelAccount()
+                    {
+                        Id = twilioEvent.From,
+                    },
+                    Recipient = new ChannelAccount()
+                    {
+                        Id = twilioEvent.To,
+                    },
+                    Text = (twilioEvent as dynamic).Body,
+                    ChannelData = twilioEvent,
+                    Type = ActivityTypes.Message,
+                };
+
+                // Detect attachments
+                if (Convert.ToInt32(twilioEvent.NumMedia) > 0)
+                {
+                    // specify a different event type for Botkit
+                    (activity.ChannelData as dynamic).botkitEventType = "picture_message";
+                }
+
+                // create a conversation reference
+                using (var context = new TurnContext(this, activity))
+                {
+                    await RunPipelineAsync(context, bot.OnTurnAsync, default);
```

Comment:
Use the cancellationToken that this method receives in the call.
Also add .ConfigureAwait(false) to the async call.